### PR TITLE
Add error handling to width inference for `invoke`

### PR DESCRIPTION
**Pull request by polybeandip**
Consider this small eDSL program
```python
# bad_invoke.py
import calyx.builder as cb

def insert_comp(prog):
    comp = prog.component("adder")

    a = comp.input("a", 32)

    comp.control += []

    return comp

def insert_main(prog):
    main = prog.component("main")

    adder = insert_comp(prog)
    adder = main.cell("adder", adder)

    main.control += [cb.invoke(adder, in_a=10)]

if __name__ == "__main__":
    prog = cb.Builder()
    insert_main(prog)
    prog.program.emit()
```
For examples like these, the eDSL is unable to automatically infer the widths of constant inputs to `invoke`: in this case`10` and `12`. Instead of throwing an error when running `python bad_invoke.py`, it generates the following Calyx code, in which the widths of `a` and `b` are `None`:
```
import "primitives/core.futil";
component main() -> () {
  cells {
    adder = adder();
  }
  wires {

  }
  control {
    seq {
      invoke adder(a=None'd10, b=None'd12)();
    }
  }
}
component adder(a: 32) -> () {
  cells {

  }
  wires {

  }
  control {
    seq {

    }
  }
}
```
Until width inference is extended to handle these cases, this short PR ensures running `python <eDSL_program_name>.py` generates an error instead of malformed Calyx code.

### CODE REVIEW

File: calyx-py/calyx/builder.py [hunk 1]
```diff
@@ -1092,14 +1092,24 @@ def invoke(cell: CellBuilder, **kwargs) -> ast.Invoke:
     The keyword arguments should have the form `in_*`, `out_*`, or `ref_*`, where
     `*` is the name of an input port, output port, or ref cell on the invoked cell.
     """
+
+    def try_infer_width(x):
+        width = cell.infer_width(x)
+        if not width:
+            raise WidthInferenceError(
+                "Cannot infer width. "
+                "Consider providing width as an argument."
```
Comment (anshumanmohan) [inline calyx-py/calyx/builder.py:1101]: Nice, thanks! I wonder if we can make the error message a little more useful, though, by excavating the names of these cells. Don't spend too much time on this, but here are a few options in increasing order of difficulty and niceness:
- "Could not infer width of input during invoke. Consider..."
- "Could not infer width of input when invoking component `adder`. Consider..."
- "Could not infer width of input `a` when invoking component `adder`. Consider..."

**Review by anshumanmohan — CHANGES_REQUESTED** [review]
Thanks for the catch and the fix. A little suggestion below to make the error message more useful.

**Comment by anshumanmohan** [comment]
Also I edited your topmatter just a hair: added `python` after your first three backticks to get nicer formatting in the code listing. Hope you don't mind, just reads easier!

File: calyx-py/calyx/builder.py (hunk 1 above)
Comment (polybeandip) [reply calyx-py/calyx/builder.py:1101]: These make total sense. I believe this last commit should construct error messages like option 3.

**Comment by anshumanmohan** [comment]
I added a quick commit to change the error message a little more.
- In truth we are inferring the name of the local component's _cell_ by which it will call the invoked component, not the invoked component itself. 
- Unlike some other parts of the eDSL, the user cannot actually pass a width as an argument, they must use `const` to specify it. I clarified that in the error message.

If these look good to you, go ahead and merge!

**Review by anshumanmohan — APPROVED** [review]

**Pull request merged**